homeassistant/components/zwave_mqtt/__init__.py

```diff
@@ -0,0 +1,332 @@
+        _LOGGER.debug("[NODE ADDED] node_id: %s", node.id)
+        data_nodes[node.id] = node
+        if node.id not in data_values:
+            data_values[node.id] = []
+
+    @callback
+    def async_node_changed(node):
+        _LOGGER.debug("[NODE CHANGED] node_id: %s", node.id)
+        data_nodes[node.id] = node
+        # notify devices about the node change
+        if node.id not in removed_nodes:
+            hass.async_create_task(handle_node_update(hass, node))
+
+    @callback
+    def async_node_removed(node):
+        _LOGGER.debug("[NODE REMOVED] node_id: %s", node.id)
+        data_nodes.pop(node.id)
+        # node added/removed events also happen on (re)starts of hass/mqtt/ozw
+        # cleanup device/entity registry if we know this node is permanently deleted
+        # entities itself are removed by the values logic
+        if node.id in removed_nodes:
+            hass.async_create_task(handle_remove_node(hass, node))
+            removed_nodes.remove(node.id)
+
+    @callback
+    def async_instance_event(message):
+        event = message["event"]
+        event_data = message["data"]
+        _LOGGER.debug("[INSTANCE EVENT]: %s - data: %s", event, event_data)
+        # The actual removal action of a Z-Wave node is reported as instance event
+        # Only when this event is detected we cleanup the device and entities from hass
+        if event == "removenode" and "Node" in event_data:
+            removed_nodes.append(event_data["Node"])
+
+    @callback
+    def async_value_added(value):
+        node = value.node
+        node_id = value.node.node_id
+
+        # Filter out CommandClasses we're definitely not interested in.
+        if value.command_class in [
+            CommandClass.CONFIGURATION,
+            CommandClass.VERSION,
+            CommandClass.MANUFACTURER_SPECIFIC,
+        ]:
+            return
+
+        _LOGGER.debug(
+            "[VALUE ADDED] node_id: %s - label: %s - value: %s - value_id: %s - CC: %s",
+            value.node.id,
+            value.label,
+            value.value,
+            value.value_id_key,
+            value.command_class,
+        )
+
+        node_data_values = data_values[node_id]
+
+        # Check if this value should be tracked by an existing entity
+        value_unique_id = f"{value.node.id}-{value.value_id_key}"
+        for values in node_data_values:
+            values.check_value(value)
+            if values.values_id == value_unique_id:
+                return  # this value already has an entity
+
+        # Run discovery on it and see if any entities need created
+        for schema in DISCOVERY_SCHEMAS:
+            if not check_node_schema(node, schema):
+                continue
+            if not check_value_schema(
+                value, schema[const.DISC_VALUES][const.DISC_PRIMARY]
+            ):
+                continue
+
+            values = ZWaveDeviceEntityValues(hass, options, schema, value)
+            values.setup()
```

Comment:
Should the `values` methods be prefixed with `async_` ?

Reply:
Commit done.